impl/src/main/java/org/jboss/weld/resolution/AbstractAssignabilityRules.java

```diff
@@ -54,7 +54,7 @@ public boolean matches(Type requiredType, Set<? extends Type> beanTypes) {
      * Standard Java covariant assignability rules are applied to all other types of bounds.
      * This is not explicitly mentioned in the specification but is implied.
      */
-    protected Type[] getUppermostTypeVariableBounds(TypeVariable<?> bound) {
+    static Type[] getUppermostTypeVariableBounds(TypeVariable<?> bound) {
```

Comment:
(That's a backwards-incompatible change, right? Is that OK?)

Reply:
For an internal Weld class, I'd say yes, this is fine.

Reply:
OK. Are there any conclusions to draw from this? i.e. anything in `impl` can change in any way without prior notice, even if it's `public`? (Does `impl` mean "internal Weld class"? Is that defined/specified/documented anywhere?)

Reply:
The class is public though? But if it's not supposed to be used or extended, not even by integrators, then I agree it's fine. Restoring compatibility should be pretty straightforward if required.

Reply:
IIRC only the `weld-api` and `weld-spi` artifacts are considered a public API in Weld. The first for users and the latter [for intergrators](https://docs.jboss.org/weld/reference/latest/en-US/html_single/#ri-spi). I don't think it's explicitly defined somewhere but the very existence of these artifacts indicates what the intention was.

Reply:
I agree with Martin, there is a reason why we have API and SPI artifacts as well.
You can technically depend on/use any internal part of Weld but those are subject to change, albeit very rarely these days.